src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -1142,6 +1188,84 @@ protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot
 
             // mark the variable's slot so that we don't complain about the variable again
             _alreadyReported[slot] = true;
+
+            bool doesNotNeedImplicitFieldInitialization()
+                => !_requireOutParamsAssigned
+                || CurrentSymbol is not MethodSymbol { MethodKind: MethodKind.Constructor, ContainingType.TypeKind: TypeKind.Struct };
+
+            ErrorCode adjustErrorForStructThis(Symbol thisParameter, int thisSlot)
+            {
+                if (doesNotNeedImplicitFieldInitialization())
+                {
+                    return ErrorCode.ERR_UseDefViolationThis;
+                }
+
+                foreach (var field in _emptyStructTypeCache.GetStructInstanceFields(thisParameter.ContainingType))
+                {
+                    if (_emptyStructTypeCache.IsEmptyStructType(field.Type))
+                        continue;
+
+                    if (field is TupleErrorFieldSymbol)
+                        continue;
+
+                    int slot = VariableSlot(field, thisSlot);
+                    if (slot == -1 || !State.IsAssigned(slot))
+                    {
+                        Debug.Assert((object)field.ContainingType == CurrentSymbol.EnclosingThisSymbol().Type);
+                        AddImplicitlyInitializedField(field);
+                    }
+                }
+
+                if (!compilation.IsFeatureEnabled(MessageID.IDS_FeatureImplicitInitializationInStructConstructors))
+                {
+                    return ErrorCode.ERR_UseDefViolationThis;
+                }
+
+                return ErrorCode.WRN_UseDefViolationStructThis;
+            }
+
+            ErrorCode adjustErrorForStructFieldOrProperty(bool isProperty, int fieldSlot)
```

Comment:
>adjustErrorForStructFieldOrProperty [](http://example.com/codeflow?start=22&length=35)

nit: I wonder if the local function could benefit from just returning a boolean, to reduce duplication of return statements: `bool downgradeErrorForStructFieldOrProperty(int fieldSlot)`. 
Then the call sites are distinct: `errorCode = downgradeErrorForStructFieldOrProperty(fieldSlot) ? ErrorCode1 : ErrorCode2;` with the relevant error codes.

Same idea applies to `adjustErrorForStructThis`.